File: bindings/rust/src/gpu.rs
//! GPU acceleration operations.

use crate::error::{check_error, Error, Result};
use crate::ffi::{self, AbiGpu, AbiGpuConfig};
use std::ffi::CStr;
use std::ptr;

/// GPU backend type.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum Backend {
    /// Auto-detect the best available backend.
    Auto = 0,
    /// NVIDIA CUDA.
    Cuda = 1,
    /// Vulkan (cross-platform).
    Vulkan = 2,
    /// Apple Metal (macOS/iOS).
    Metal = 3,
    /// WebGPU (browser/WASM).
    WebGpu = 4,
}

impl Default for Backend {
    fn default() -> Self {
        Backend::Auto
    }
}

/// GPU context for acceleration.
///
/// # Example
///
/// ```no_run
/// use abi::Gpu;
///
/// if Gpu::is_available() {
///     let gpu = Gpu::new().expect("Failed to initialize GPU");
///     println!("Using backend: {}", gpu.backend_name());
/// }
/// ```
pub struct Gpu {
    handle: AbiGpu,
}

impl Gpu {
    /// Initialize GPU context with auto-detected backend.
    pub fn new() -> Result<Self> {
        Self::with_config(GpuConfig::default())
    }

    /// Initialize GPU context with specific configuration.
    pub fn with_config(config: GpuConfig) -> Result<Self> {
        let ffi_config = AbiGpuConfig {
            backend: config.backend as i32,
            device_index: config.device_index as i32,
            enable_profiling: config.enable_profiling,
        };

        let mut handle: AbiGpu = ptr::null_mut();
        let err = unsafe { ffi::abi_gpu_init(&ffi_config, &mut handle) };
        check_error(err)?;

        if handle.is_null() {
            return Err(Error::NullPointer);
        }

        Ok(Self { handle })
    }

    /// Check if any GPU backend is available.
    pub fn is_available() -> bool {
        unsafe { ffi::abi_gpu_is_available() }
    }

    /// Get the active backend name.
    pub fn backend_name(&self) -> &str {
        unsafe {
            let ptr = ffi::abi_gpu_backend_name(self.handle);
            if ptr.is_null() {
                "unknown"
            } else {
                CStr::from_ptr(ptr).to_str().unwrap_or("unknown")
            }
        }
    }
}

impl Drop for Gpu {
    fn drop(&mut self) {
        if !self.handle.is_null() {
            unsafe {
                ffi::abi_gpu_shutdown(self.handle);
            }
        }
    }
}

unsafe impl Send for Gpu {}

/// GPU configuration.
#[derive(Debug, Clone)]
pub struct GpuConfig {
    /// Backend to use.
    pub backend: Backend,
    /// Device index (0 = first GPU).
    pub device_index: u32,
    /// Enable profiling.
    pub enable_profiling: bool,
}

impl Default for GpuConfig {
    fn default() -> Self {
        Self {
            backend: Backend::Auto,
            device_index: 0,
            enable_profiling: false,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_gpu_available() {
        // This should work even without the library
        let _ = Gpu::is_available();
    }
}

File: bindings/c/src/gpu.zig
//! C-compatible GPU operation exports.
//! Provides GPU context management for FFI.

const std = @import("std");
const builtin = @import("builtin");
const errors = @import("errors.zig");

/// Opaque GPU handle for C API.
pub const GpuHandle = opaque {};

/// GPU config matching C header (abi_gpu_config_t).
pub const GpuConfig = extern struct {
    backend: c_int = 0, // 0=auto, 1=cuda, 2=vulkan, 3=metal, 4=webgpu
    device_index: c_int = 0,
    enable_profiling: bool = false,
};

/// GPU backend enum.
pub const Backend = enum(c_int) {
    auto = 0,
    cuda = 1,
    vulkan = 2,
    metal = 3,
    webgpu = 4,
    cpu = 5,
};

/// Internal GPU state.
const GpuState = struct {
    allocator: std.mem.Allocator,
    backend: Backend,
    device_index: c_int,
    profiling_enabled: bool,

    pub fn init(allocator: std.mem.Allocator, config: GpuConfig) !*GpuState {
        const state = try allocator.create(GpuState);
        errdefer allocator.destroy(state);

        // Detect best backend
        const backend = detectBackend(@enumFromInt(config.backend));

        state.* = .{
            .allocator = allocator,
            .backend = backend,
            .device_index = config.device_index,
            .profiling_enabled = config.enable_profiling,
        };

        return state;
    }

    pub fn deinit(self: *GpuState) void {
        self.allocator.destroy(self);
    }

    pub fn backendName(self: *const GpuState) [*:0]const u8 {
        return switch (self.backend) {
            .auto => "auto",
            .cuda => "cuda",
            .vulkan => "vulkan",
            .metal => "metal",
            .webgpu => "webgpu",
            .cpu => "cpu",
        };
    }
};

/// Detect best available backend.
fn detectBackend(requested: Backend) Backend {
    if (requested != .auto) {
        return requested;
    }

    // Auto-detect based on platform
    const os = builtin.os.tag;

    if (os == .macos) {
        return .metal;
    }

    // Default to CPU fallback
    return .cpu;
}

/// Check if any GPU is potentially available.
fn isGpuAvailable() bool {
    const os = builtin.os.tag;

    // Metal available on macOS
    if (os == .macos) return true;

    // Other platforms might have Vulkan or CUDA
    if (os == .linux or os == .windows) return true;

    return false;
}

// Global allocator
var gpa = std.heap.GeneralPurposeAllocator(.{}){};

/// Initialize GPU context.
pub export fn abi_gpu_init(config: *const GpuConfig, out_gpu: *?*GpuHandle) errors.Error {
    const allocator = gpa.allocator();
    const state = GpuState.init(allocator, config.*) catch |err| {
        return errors.fromZigError(err);
    };
    out_gpu.* = @ptrCast(state);
    return errors.OK;
}

/// Shutdown GPU context.
pub export fn abi_gpu_shutdown(handle: ?*GpuHandle) void {
    if (handle) |h| {
        const state: *GpuState = @ptrCast(@alignCast(h));
        state.deinit();
    }
}

/// Check if any GPU backend is available.
pub export fn abi_gpu_is_available() bool {
    return isGpuAvailable();
}

/// Get active GPU backend name.
pub export fn abi_gpu_backend_name(handle: ?*GpuHandle) [*:0]const u8 {
    if (handle) |h| {
        const state: *const GpuState = @ptrCast(@alignCast(h));
        return state.backendName();
    }
    return "none";
}

test "gpu exports" {
    var gpu: ?*GpuHandle = null;
    const config = GpuConfig{
        .backend = 0, // auto
        .device_index = 0,
        .enable_profiling = false,
    };

    try std.testing.expectEqual(errors.OK, abi_gpu_init(&config, &gpu));
    try std.testing.expect(gpu != null);

    const name = abi_gpu_backend_name(gpu);
    try std.testing.expect(std.mem.len(name) > 0);

    _ = abi_gpu_is_available();

    abi_gpu_shutdown(gpu);

    // Null handle should return "none"
    try std.testing.expectEqualStrings("none", std.mem.span(abi_gpu_backend_name(null)));
}
